src/raftstore/store/worker/read.rs

```diff
@@ -0,0 +1,884 @@
+            leader_lease: None,
+            tag: format!("[region {}] {}", region_id, peer_id),
+        }
+    }
+
+    fn update(&mut self, progress: Progress) {
+        if let Some(region) = progress.region {
+            self.region = region;
+        }
+        if let Some(term) = progress.term {
+            self.term = term;
+        }
+        if let Some(applied_index_term) = progress.applied_index_term {
+            self.applied_index_term = applied_index_term;
+        }
+        if let Some(lease) = progress.leader_lease {
+            self.leader_lease = Some(lease);
+        }
+    }
+
+    fn handle_read(&self, req: &RaftCmdRequest, kv_engine: &Arc<DB>) -> Option<ReadResponse> {
+        if let Ok(mut resp) =
+            ReadExecutor::new(&self.region, kv_engine, &self.tag).execute(req, false)
+        {
+            if let Some(ref lease) = self.leader_lease {
+                if lease.inspect(None) == LeaseState::Valid {
+                    // Leader can read local if and only if it is in lease.
+                    cmd_resp::bind_term(&mut resp.response, self.term);
+                    return Some(resp);
+                }
+            }
+        }
+        None
+    }
+}
+
+impl Display for ReadDelegate {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "ReadDelegate for region {}, \
+             leader {} at term {}, applied_index_term {}, has lease {}",
+            self.region.get_id(),
+            self.peer_id,
+            self.term,
+            self.applied_index_term,
+            self.leader_lease.is_some(),
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct Progress {
+    region: Option<metapb::Region>,
+    term: Option<u64>,
+    applied_index_term: Option<u64>,
+    leader_lease: Option<RemoteLease>,
+}
+
+impl Progress {
+    pub fn new() -> Self {
+        Progress {
+            region: None,
+            term: None,
+            applied_index_term: None,
+            leader_lease: None,
+        }
+    }
+
+    pub fn set_region(&mut self, region: metapb::Region) {
+        self.region = Some(region);
+    }
+
+    pub fn set_term(&mut self, term: u64) {
+        self.term = Some(term);
+    }
+
+    pub fn set_applied_index_term(&mut self, applied_index_term: u64) {
+        self.applied_index_term = Some(applied_index_term);
+    }
+
+    pub fn set_leader_lease(&mut self, lease: RemoteLease) {
+        self.leader_lease = Some(lease);
+    }
+}
+
+pub enum Task {
+    Register(ReadDelegate),
+    Update((u64, Progress)),
+    Read(StoreMsg),
+    Destroy(u64),
+}
+
+impl Task {
+    pub fn register(peer: &Peer) -> Task {
+        let delegate = ReadDelegate::from_peer(peer);
+        Task::Register(delegate)
+    }
+
+    pub fn update(region_id: u64, progress: Progress) -> Task {
+        Task::Update((region_id, progress))
+    }
+
+    pub fn destroy(region_id: u64) -> Task {
+        Task::Destroy(region_id)
+    }
+
+    #[inline]
```

Comment:
Please add documentation to this function. It's not clear why this is part of `Task` or what exactly it is communicating.